validation-test/stdlib/Set.swift

```diff
@@ -2861,67 +2861,83 @@ SetTestSuite.test("isSubsetOf.Set.Set") {
   let s2 = Set([1010, 2020, 3030])
   expectTrue(Set<Int>().isSubset(of: s1))
   expectFalse(s1.isSubset(of: Set<Int>()))
+  expectTrue(Set<Int>().isSubset(of: Set<Int>()))
   expectTrue(s1.isSubset(of: s1))
+  expectFalse(s1.isSubset(of: s2))
   expectTrue(s2.isSubset(of: s1))
 }
 
 SetTestSuite.test("isSubsetOf.Set.Sequence") {
-  let s1 = Set([1010, 2020, 3030, 4040, 5050, 6060])
-  expectTrue(Set<Int>().isSubset(of: s1))
-  expectFalse(s1.isSubset(of: Set<Int>()))
-  expectTrue(s1.isSubset(of: s1))
+  let s1 = Set([1010, 2020, 3030])
+  expectTrue(Set<Int>().isSubset(of: [1010]))
+  expectFalse(s1.isSubset(of: []))
+  expectTrue(Set<Int>().isSubset(of: []))
+  expectTrue(s1.isSubset(of: [1010, 2020, 3030]))
```

Comment:
The array literals are a little misleading here: Set and Array are both expressible by such, and it seems easy to misread this as a Set-Set operation.

I liked how some of the original code used explicit AnySequence constants — can we switch all these to use that form instead? (Or, alternatively, we could simply add explicit `as Array<Int>` annotations to make the meaning clear to a distracted reader.)

Reply:
I see your point, though I worry that wrapping `AnySequence` around all the arrays obscures the meaning quite a bit. In literal form it’s easiest to deduce the tests intent to systematically validate:
* same (content) sequence
* sub-sequence
* super-sequence 

I’d prefer adding the annotations to the end of literals if you’re sure it’s really clarifying here. 

I have also experimented with more explicitly named formulation, declaring all the variables with names like:
* `set`, `subSet`, `emptySet`,
* `seq`, `subSeq`, `emptySeq`, `superSeq`

Then the validation part reads much like a set of mathematical rules, but the trade-off was much longer setup part which seemed a bit foreign compared to other unit tests. What do you think about that idea?

Reply:
I'm starting to hate all these alternatives for obscuring the test intent with noisy syntactic boilerplate… 
#### `AnySequence` Wrapper
````swift
SetTestSuite.test("isSubsetOf.Set.Sequence") {
  let s1 = Set([1010, 2020, 3030])
  expectTrue(Set<Int>().isSubset(of: AnySequence([1010])))
  expectFalse(s1.isSubset(of: AnySequence([])))
  expectTrue(Set<Int>().isSubset(of: AnySequence([])))
  expectTrue(s1.isSubset(of: AnySequence([1010, 2020, 3030])))
  expectFalse(s1.isSubset(of: AnySequence([1010, 2020])))
  expectTrue(s1.isSubset(of: AnySequence([1010, 2020, 3030, 4040])))
}
````
#### Explicit Type Cast Annotation `as Array<Int>`
````swift
SetTestSuite.test("isSubsetOf.Set.Sequence") {
  let s1 = Set([1010, 2020, 3030])
  expectTrue(Set<Int>().isSubset(of: [1010] as Array<Int>))
  expectFalse(s1.isSubset(of: [] as Array<Int>))
  expectTrue(Set<Int>().isSubset(of: [] as Array<Int>))
  expectTrue(s1.isSubset(of: [1010, 2020, 3030] as Array<Int>))
  expectFalse(s1.isSubset(of: [1010, 2020] as Array<Int>))
  expectTrue(s1.isSubset(of: [1010, 2020, 3030, 4040] as Array<Int>))
}
````
Even the **Semantic Variable Names** approach I proposed above isn't much better...
````swift
SetTestSuite.test("isSubsetOf.Set.Sequence") {
  let aSet = Set([1010, 2020, 3030])
  let emptySet = Set<Int>()
  let aSeq = [1010, 2020, 3030]
  let subSeq = [1010, 2020]
  let superSeq = [1010, 2020, 3030, 4040]
  let emptySeq = Array<Int>()
  expectTrue(emptySet.isSubset(of: aSeq))
  expectFalse(aSet.isSubset(of: emptySeq))
  expectTrue(emptySet.isSubset(of: emptySeq))
  expectTrue(aSet.isSubset(of: aSeq))
  expectFalse(aSet.isSubset(of: subSeq))
  expectTrue(aSet.isSubset(of: superSeq))
}
````
…unless we extract the `Set` and `Sequence` declarations in some kind of a share preamble for these tests. Otherwise they double the test length which, again, obscures more than clarifies. But that goes against the self contained style used in all the other unit tests.

You said that (emphasis mine):
> The array literals are a little misleading here: Set and Array are both expressible by such, and it seems **easy to misread this as a Set-Set** operation.

But ultimately, I find it implausible that these unit tests can be misread this way, given that:
* the declaration clearly labels it as `"operation.Set.Sequence"`,
* the `Set.Set` and `Set.Sequence` variants are always interleaved in that order,
* both variants fit on the screen at the same time, next to each other
(allowing for easy verification of them being the same logical test).

The only risk I see, is that the tests written with array literals are currently wrong due to type inference and implicit conversions. Because if the array literal to `Set` conversion would have higher priority than the array literal to `Array<Int>` via `Sequence` conformance, it would indeed be a disastrous pitfall in the language itself. I hope there are validation tests that verify this isn't the case elsewhere in the test suite. Right?

If this is a real risk, than I propose we simplify the type annotation using type alias (removing the repeated generic noise):

````swift
typealias Seq = Array<Int>

SetTestSuite.test("isSubsetOf.Set.Sequence") {
  let s1 = Set([1010, 2020, 3030])
  expectTrue(Set<Int>().isSubset(of: [1010] as Seq))
  expectFalse(s1.isSubset(of: [] as Seq))
  expectTrue(Set<Int>().isSubset(of: [] as Seq))
  expectTrue(s1.isSubset(of: [1010, 2020, 3030] as Seq))
  expectFalse(s1.isSubset(of: [1010, 2020] as Seq))
  expectTrue(s1.isSubset(of: [1010, 2020, 3030, 4040] as Seq))
}
````

@lorentey, @airspeedswift, @dabrahams please pick your poison.

Reply:
Thank you all for participating in my rubber duck debugging session. Typealias it is!